subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingClasspathFileTransformer.java

```diff
@@ -67,12 +67,22 @@ public File transform(File source, CompleteFileSystemLocationSnapshot sourceSnap
         HashCode fileHash = hasher.hash();
         File transformed = new File(cacheDir, fileHash.toString() + '/' + name);
         if (!transformed.isFile()) {
-            transform(source, transformed);
+            transformed = transform(source, transformed, sourceSnapshot, cacheDir);
         }
         return transformed;
     }
 
-    private void transform(File source, File dest) {
+    private File transform(File source, File dest, CompleteFileSystemLocationSnapshot sourceSnapshot, File cacheDir) {
+        if (isSignedJar(source)) {
+            LOGGER.debug("Signed archive '{}'. Skipping instrumentation.", source.getName());
+            return copyingClasspathFileTransformer.transform(source, sourceSnapshot, cacheDir);
```

Comment:
This has to guarantee that the file ends up at exactly the `dest` location, otherwise the file will be treated as missing next time we look up in the cache and we'll run through the whole process of opening the jar and checking for signature files again. Delegating to the other transformer here is just a bit too coincidental and too easy to break this contract. So, I would explicitly do the copy here instead.

Reply:
Good catch, fixed